tests/handlers/auth_test.py

```diff
@@ -0,0 +1,358 @@
+    assert header_method == method
+    if header_method == "Basic":
+        assert header_info == 'realm="tokens"'
+    else:
+        data = header_info.split(",")
+        assert data[0] == 'realm="tokens"'
+        assert data[1] == f'error="{error}"'
+        assert data[2].startswith("error_description=")
+
+
+async def test_authnz_token_no_auth(aiohttp_client: TestClient) -> None:
+    app = await create_test_app()
+    client = await aiohttp_client(app)
+
+    r = await client.get("/auth", params={"capability": "exec:admin"})
+    assert r.status == 401
+    assert r.headers["WWW-Authenticate"]
+    assert_www_authenticate_header_matches(
+        r.headers["WWW-Authenticate"], "Bearer", "No Authorization header"
+    )
+
+    r = await client.get(
+        "/auth",
+        params={"capability": "exec:admin"},
+        headers={"Authorization": "Bearer"},
+    )
+    assert r.status == 401
+    assert r.headers["WWW-Authenticate"]
+    assert_www_authenticate_header_matches(
+        r.headers["WWW-Authenticate"], "Bearer", "Unable to find token"
+    )
+
+    r = await client.get(
+        "/auth",
+        params={"capability": "exec:admin"},
+        headers={"Authorization": "Bearer token"},
+    )
+    assert r.status == 401
+    assert r.headers["WWW-Authenticate"]
+    assert_www_authenticate_header_matches(
+        r.headers["WWW-Authenticate"], "Bearer", "Invalid token"
+    )
+
+
+async def test_authnz_token_access_denied(aiohttp_client: TestClient) -> None:
+    keypair = RSAKeyPair()
+    token = create_test_token(keypair)
+    app = await create_test_app(keypair)
+    client = await aiohttp_client(app)
+
+    r = await client.get(
+        "/auth",
+        params={"capability": "exec:admin"},
+        headers={"Authorization": f"Bearer {token}"},
+    )
+    assert r.status == 403
+    body = await r.text()
+    assert "No Capability group found in user's `isMemberOf`" in body
+    assert r.headers["X-Auth-Request-Token-Capabilities"] == ""
+    assert r.headers["X-Auth-Request-Capabilities-Accepted"] == "exec:admin"
+    assert r.headers["X-Auth-Request-Capabilities-Satisfy"] == "all"
+
+
+async def test_authnz_token_satisfy_all(aiohttp_client: TestClient) -> None:
+    keypair = RSAKeyPair()
+    token = create_test_token(keypair, ["test"])
+    app = await create_test_app(keypair)
+    client = await aiohttp_client(app)
+
+    r = await client.get(
+        "/auth",
+        params=[("capability", "exec:test"), ("capability", "exec:admin")],
+        headers={"Authorization": f"Bearer {token}"},
+    )
+    assert r.status == 403
+    body = await r.text()
+    assert "No Capability group found in user's `isMemberOf`" in body
+    assert r.headers["X-Auth-Request-Token-Capabilities"] == "exec:test"
+    assert (
+        r.headers["X-Auth-Request-Capabilities-Accepted"]
+        == "exec:admin exec:test"
+    )
+    assert r.headers["X-Auth-Request-Capabilities-Satisfy"] == "all"
```

Comment:
This is probably scope creep, but just thinking out loud, it might be useful to try to collect logs of issues with capabilities if we decide to go that way.  That might make it more obvious when your jobs or things are going wrong or lack the proper capabilities when the user might not be directly connected to the thing doing the work and getting the error.  Just a thought.

Reply:
Yes. I definitely want to do that -- my tentative thought is to start reporting all authentication decisions via Kafka so that we can do statistics and easily see what's going on, although it may be easier to do it with structured logging and ELK instead.